doc/gntgaim.1.in
.\" Copyright (c) 2006, Sadrul Habib Chowdhury <sadrul@users.sf.net>
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139,
.\" USA.
.TH gntgaim 1
.SH NAME
GntGaim \- A Pimpin' Penguin console frontend for Instant Messaging client Gaim.
.SH SYNOPSIS
.TP 5
\fBgntgaim \fI[options]\fR

.SH DESCRIPTION
.PP
\fBgntgaim\fR is a console-based modular messaging client capable of using
AIM, MSN, Yahoo!, Jabber, ICQ, IRC, SILC, Novell GroupWise, Lotus Sametime,
Zephyr, Gadu-Gadu, and QQ all at once. It has many common features found in
other clients, as well as many unique features.  Gaim is not endorsed by or
affiliated with America Online, ICQ, Microsoft, or Yahoo.

.SH OPTIONS
The following options are provided by \fBgntgaim\fR using the standard GNU
command line syntax:
.TP
.B \-d, \-\-debug
Print debugging messages to stderr and start with the \fBDebug\fR window. The
messages shown in the \fBDebug\fR window are the same as the ones printed in
stderr.
.TP
.B \-c, \-\-config=\fIDIR\fB
Use \fIDIR\fR as the directory for config files instead of \fI~/.gaim\fR.
.TP
.B \-h, \-\-help
Print this help and exit.
.TP
.B \-n, \-\-nologin
Don't automatically login when \fBgntgaim\fR starts.  Sets all accounts to
Offline.
.TP
.B \-v, \-\-version
Display the version information window.

.SH GNT Shortcuts
You can use the following shortcuts:
.TP
.B Alt \+ n
Go to the next window.
.TP
.B Alt \+ p
Go to the previous window.
.TP
.B Alt \+ w
Show the list of windows. You can select and jump to any window from the list.
.TP
.B Alt \+ c
Close the current window.
.TP
.B Alt \+ q
Quit.
.TP
.B Alt \+ m
Start moving a window. Press the cursor keys to move the window. When you are
done, press \fBEnter\fR or \fBEscape\fR.
.TP
.B Alt \+ r
Start resizing a window. Press the cursor keys to resize the window. When you
are done, press \fBEnter\fR or \fBEscape\fR.
.TP
.B Alt \+ d
Dump the contents of the screen in HTML format in a file named "dump.html" in
working directory.
.TP
.B Alt \+ .
Move the position of the current window in the window list one place to the
right.
.TP
.B Alt \+ ,
Move the position of the current window in the window list one place to the
left.
.TP
.B Alt \+ l
Refresh the windows. This is useful after resizing the terminal window.
.TP
.B Alt \+ 1 2 ... 0
Jump to the 1st, 2nd ... 10th window.

.SH FILES
\fI~/.gntrc\fR: configuration file for gnt applications.
.br
.TP
A sample file looks like:
.br
[general]
.br
shadow = 0
.br
# There is experimental mouse support
.br
mouse = 1
.br
# To use some custom window-manager
.br
wm = /usr/local/lib/gaim/s.so
.br

.br
[colors]
.br
# The RGB values range in [0, 1000]
.br
black = 0; 0; 0
.br
red = 1000; 0; 0
.br
green = 0; 1000; 0
.br
blue = 250; 250; 700
.br
white = 1000; 1000; 1000
.br
gray = 700; 700; 700
.br
darkgray = 256; 256; 256
.br

.br
[colorpairs]
.br
normal = white; black
.br
highlight = white; blue
.br
highlightd = black; gray
.br
shadow = black; darkgray
.br
title = white; blue
.br
titled = white; gray
.br
text = white; blue
.br
disabled = gray; black
.br

.br
# Remap some keys for GntEntry
.br
[GntEntry::remap]
.br
# Remap the up-arrow to the left-arrow
.br
^[[A = ^[[D
.br
# Remap the down-arrow to the right-arrow
.br
^[[B = ^[[C
.br
# Remap 'a' to 'bcd'
.br
a = bcd
.br
# Completely ignore the key 'q'
.br
q = 
.br
# But the following will NOT work
.br
#abc = bcd
.br

# Hitting 'space' will activate a button
.br
[GntButton::remap]
.br
\\  = \\r
.br

.SH Mouse Support
There is experimental mouse support. You can focus windows, activate buttons,
select rows in a list, scroll using the wheel-scroll etc. Mouse support is
disabled by default, so you need to enable it in ~/.gntrc (see the sample
above).

.SH Window Management
The default window management is very limited. But it is possible to write
custom window managers to suit your needs. There is a sample window-manager
included (named \fIs.so\fR) which adds a little 'close-button' for the windows,
removes the borders from the buddylist and pops up new windows in the middle of
the screen, instead of at the upper-left corder. It is provided as a sample
simple manager, but it should be possible to write more complex managers, and
it's very possible that the window-manager API will need to be enhanced. Look at
the sample ~/.gntrc file above to see how to select a window-manager.


.SH Conversation Commands
There are a few helpful commands in addition to the regular commands. You can
use these from any conversation to access other windows.

.TP
.B accounts
for the accounts window.

.TP
.B buddylist
for the buddylist.

.TP
.B debugwin
for the debug window.

.TP
.B plugins
for the plugins window.

.TP
.B prefs
for the preferences window.

.TP
.B status
for the status window.

.SH BUGS
Known bugs are listed at
.br
\fIhttp://sourceforge.net/tracker/?group_id=235&atid=100235\fR

.SH PATCHES
If you fix a bug in \fBgntgaim\fR (or otherwise enhance it), please submit a
patch (using \fIsvn diff > my.diff\fR against the latest SVN version)
at
.br
\fIhttp://sourceforge.net/tracker/?func=add&group_id=235&atid=300235\fR

Before sending a bug report, please verify that you have the latest
version of \fBgntgaim\fR and \fBlibgaim\fR.  Many bugs (major and minor) are
fixed at each release, and if yours is out of date, the problem may already have
been solved.

You are also encouraged to drop by at \fB#gaim\fR on \fIirc.freenode.net\fR to
discuss about development.

.SH SEE ALSO
\fIhttp://gaim.sourceforge.net/\fR
.br
\fIhttp://www.sourceforge.net/projects/gaim/\fR

.SH LICENSE
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful, but
\fBWITHOUT ANY WARRANTY\fR; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

.SH AUTHORS
Sadrul Habib Chowdhury <\fIsadrul@users.sourceforge.net\fR>
.br

This manpage was written by Sadrul Habib Chowdhury
<\fIsadrul@users.sourceforge.net\fR> and Dennis Ristuccia
<\fIdennis@dennisr.net\fR>.



NEWS
-=[ Gaim ]=-  The Pimpin' Penguin IM Client That's Good For The Soul!

2.0.0beta3 (03/25/2006):
	Mark: Yeah, I know, another beta.  Don't worry, we'll get this
	puppy out the door eventually.

2.0.0beta2 (01/24/2006):
	Mark: So this is the new year, and I don't feel any different, but
	Gaim is getting better.  We hope this will be our last beta before
	we release the final version of 2.0.0.  As before, please shower us
	in feedback!

	Richard: I'm proud to say a lot of bugs have been squashed in this
	version.  If you filed a bug against beta1, please test to see if
	it's fixed now and update your bug report accordingly (by either
	closing it or setting the version to 2.0.0beta2).  I'm also looking
	for someone who uses Gaim on MacOS X to test a patch for me before
	I can commit it.  See http://gaim.sf.net/contactinfo.php for my
	contact information.

	Etan: So I did a bunch of work on the perl plugin since beta1, so
	anyone who uses perl plugins would do well to expect some things to
	need updating (I'm not certain everything works yet, so please send me
	any reports of things that don't). Most of the work was correcting
	some namespace issues, but I also improved the support for perl
	plugins having plugin_pref frames, and plugin actions. Multiple perl
	plugins can now have plugin_pref frames at the same time, and every
	perl plugin can have multiple plugin actions now. Like Mark said
	above, let us know how this beta works out.

2.0.0beta1 (12/17/2005):
	Sean: I think Nathan sums everything up really well below. There's still
	a bunch we want to add (and remove) before the official release, but we
	really want to start getting feedback about what's good and what's not.
	So, please, be vocal about this beta!

	Nathan: 15 months since we branched oldstatus, and started working on
	the behemoth that is 2.0.0.  In that time, we've added a couple new
	protocols, we had a few crazy patch writers become developers, and
	had a few more people step up to be crazy patch writers.  Sean wrote
	a book, and we arbitrarily decided to make a version 1.0.0, and a
	new versioning scheme.  We got new artwork, and added almost 200 lines
	to the ChangeLog.  We've watched 2 major GTK+ releases, and added all
	kinds of features using them.  OK, I've wasted enough bits here.  This
	beta rocks, but it is a beta.  Treat it as such, and enjoy!

	Gary: Finally we have a beta.  There are a lot of new goodies, including
	quite a few summer of code projects that couldn't get added into the
	oldstatus branch because API changes.  We also have a new mono plugin
	loader for even more plugin fun.

	Richard: Thanks to all who wrote patches (big or small) for this beta.

	Tim: Well here it is, the first beta. There's a lot of cool things in
	here, but not all of them are finished or debugged. But then that's
	why it's called a beta. Conversations are now contact-aware, and
	there's a new status selector. There's smooth scrolling on incoming
	messages as well. Our Crazy Patch Writers have been doing a good
	job too, it seems like we're never lacking some patches in the
	tracker to review.

	Etan: I know I'm not going to be able to think of all the things I should
	talk about here so I'll just go with the stuff I remember. Adding buddies
	on ejabberd jabber servers should work more correctly now (it's possible
	adding buddies on other servers is a bit broken currently I'm still
	looking into it). There are still other jabber issues I'm looking into
	with handling of buddies.
	In other topics, I finally brought my Accounts menu into gaim, though the
	current example of it isn't exactly what I had in mind, anyone with any
	suggestions/comments/etc about it please speak up.
	Other than the stuff I'm sure I'm forgetting to mention this beta should
	be awesome as gaim 2.0.0 is going to be awesome. Have fun with it.

	Luke: pretty much everything important has been said, so this is going to
	be rather repetative.  This is a beta folks, so it will have bugs.  It
	will crash unexpectedly.  Perhaps even frequently.  Use it at your peril.
	But do use it (though you might want to back up ~/.gaim first), and let us
	know what needs to be fixed.

	Mark: Yay.

	Stu: Wow!

	Siege: Sametime accounts created with gaim-meanwhile should merge over
	for use in the beta just fine. Some of the familiar settings regarding
	the buddy list are gone, so I recommend backing up your Sametime blist
	before getting down and crazy. Have fun, and happy early Decemberween!

	Daniel: Enjoy! In the hopes that this prevents someone pain... do not use
	Glib/GTK+ 2.8.x with Wingaim - it will not work.  We're looking into the
	problem and hope to have it resolved before the final release.
	"Happy Holidays!"

1.5.0 (8/11/2005):
	Mark: No super crazy major changes here.  Just the usual bug
	fixes and some pretty important security updates.

	Stu: Buy Sean's book. He obviously needs the money. Other than that
	we fixed some bugs with this release (I don't think we did much else).
	SoC students are doing lots of cool stuff which we'll hopefully be
	able to bring you in a future release.

	Nathan: I've continued my streak of doing nothing useful lately.
	However, I've now got a brand-spanking-new DSL line, so I might
	actually get to contribute soon.  One way or another I'd like
	to get HEAD into a more useable state in the next month or two.
	Also, like Stu said, buy Sean's book.  I've been reading and
	fixing it for the last umpteen months, so you had better enjoy
	it.

1.4.0 (7/7/2005):
	Mark: The last month or four we've promoted a bunch of the Gaim
	Crazy Patch Writers to developers, so there is now an even larger
	team of brilliant and amazingly sexy committers working around
	the globe for your instant messaging pleasure.  Also, we have
	what I believe to be our first contribution from a Summer of Code
	student in this release: Jonathan Clark enabled the sending of
	files to certain ICQ users.  Support is still a bit rough, but

console/gntgaim.c
#include "account.h"
#include "conversation.h"
#include "core.h"
#include "debug.h"
#include "eventloop.h"
#include "ft.h"
#include "log.h"
#include "notify.h"
#include "prefs.h"
#include "prpl.h"
#include "pounce.h"
#include "savedstatuses.h"
#include "sound.h"
#include "status.h"
#include "util.h"
#include "whiteboard.h"

#include "gntdebug.h"
#include "gntgaim.h"
#include "gntprefs.h"
#include "gntui.h"

#define _GNU_SOURCE
#include <getopt.h>

#include "config.h"

static void
debug_init()
{
	gg_debug_init();
	gaim_debug_set_ui_ops(gg_debug_get_ui_ops());
}

static GaimCoreUiOps core_ops =
{
	gg_prefs_init,
	debug_init,
	NULL, /*gaim_gtk_ui_init,*/
	NULL, /*gaim_gtk_quit*/
};

static GaimCoreUiOps *
gnt_core_get_ui_ops()
{
	return &core_ops;
}

/* Anything IO-related is directly copied from gtkgaim's source tree */

#define GAIM_GTK_READ_COND  (G_IO_IN | G_IO_HUP | G_IO_ERR)
#define GAIM_GTK_WRITE_COND (G_IO_OUT | G_IO_HUP | G_IO_ERR | G_IO_NVAL)

typedef struct _GaimGtkIOClosure {
	GaimInputFunction function;
	guint result;
	gpointer data;

} GaimGtkIOClosure;

static void gaim_gtk_io_destroy(gpointer data)
{
	g_free(data);
}

static gboolean gaim_gtk_io_invoke(GIOChannel *source, GIOCondition condition, gpointer data)
{
	GaimGtkIOClosure *closure = data;
	GaimInputCondition gaim_cond = 0;

	if (condition & GAIM_GTK_READ_COND)
		gaim_cond |= GAIM_INPUT_READ;
	if (condition & GAIM_GTK_WRITE_COND)
		gaim_cond |= GAIM_INPUT_WRITE;

#if 0
	gaim_debug(GAIM_DEBUG_MISC, "gtk_eventloop",
			   "CLOSURE: callback for %d, fd is %d\n",
			   closure->result, g_io_channel_unix_get_fd(source));
#endif

#ifdef _WIN32
	if(! gaim_cond) {
#if DEBUG
		gaim_debug(GAIM_DEBUG_MISC, "gtk_eventloop",
			   "CLOSURE received GIOCondition of 0x%x, which does not"
			   " match 0x%x (READ) or 0x%x (WRITE)\n",
			   condition, GAIM_GTK_READ_COND, GAIM_GTK_WRITE_COND);
#endif /* DEBUG */

		return TRUE;
	}
#endif /* _WIN32 */

	closure->function(closure->data, g_io_channel_unix_get_fd(source),
			  gaim_cond);

	return TRUE;
}

static guint gnt_input_add(gint fd, GaimInputCondition condition, GaimInputFunction function,
							   gpointer data)
{
	GaimGtkIOClosure *closure = g_new0(GaimGtkIOClosure, 1);
	GIOChannel *channel;
	GIOCondition cond = 0;

	closure->function = function;
	closure->data = data;

	if (condition & GAIM_INPUT_READ)
		cond |= GAIM_GTK_READ_COND;
	if (condition & GAIM_INPUT_WRITE)
		cond |= GAIM_GTK_WRITE_COND;

	channel = g_io_channel_unix_new(fd);
	closure->result = g_io_add_watch_full(channel, G_PRIORITY_DEFAULT, cond,
					      gaim_gtk_io_invoke, closure, gaim_gtk_io_destroy);

	g_io_channel_unref(channel);
	return closure->result;
}

static GaimEventLoopUiOps eventloop_ops =
{
	g_timeout_add,
	(guint (*)(guint))g_source_remove,
	gnt_input_add,
	(guint (*)(guint))g_source_remove
};

static GaimEventLoopUiOps *
gnt_eventloop_get_ui_ops(void)
{
	return &eventloop_ops;
}

/* This is mostly copied from gtkgaim's source tree */
static void
show_usage(const char *name, gboolean terse)
{
	char *text;

	if (terse) {
		text = g_strdup_printf(_("%s. Try `%s -h' for more information.\n"), VERSION, name);
	} else {
		text = g_strdup_printf(_("%s\n"
		       "Usage: %s [OPTION]...\n\n"
		       "  -c, --config=DIR    use DIR for config files\n"
		       "  -d, --debug         print debugging messages to stdout\n"
		       "  -h, --help          display this help and exit\n"
		       "  -n, --nologin       don't automatically login\n"
		       "  -v, --version       display the current version and exit\n"), VERSION, name);
	}

	gaim_print_utf8_to_console(stdout, text);
	g_free(text);
}

static int
init_libgaim(int argc, char **argv)
{
	char *path;
	int opt;
	gboolean opt_help = FALSE;
	gboolean opt_nologin = FALSE;
	gboolean opt_version = FALSE;
	char *opt_config_dir_arg = NULL;
	char *opt_session_arg = NULL;
	gboolean debug_enabled = FALSE;

	struct option long_options[] = {
		{"config",   required_argument, NULL, 'c'},
		{"debug",    no_argument,       NULL, 'd'},
		{"help",     no_argument,       NULL, 'h'},
		{"nologin",  no_argument,       NULL, 'n'},
		{"session",  required_argument, NULL, 's'},
		{"version",  no_argument,       NULL, 'v'},
		{0, 0, 0, 0}
	};

	/* scan command-line options */
	opterr = 1;
	while ((opt = getopt_long(argc, argv,
#ifndef _WIN32
				  "c:dhn::s:v",
#else
				  "c:dhn::v",
#endif
				  long_options, NULL)) != -1) {
		switch (opt) {
		case 'c':	/* config dir */
			g_free(opt_config_dir_arg);
			opt_config_dir_arg = g_strdup(optarg);
			break;
		case 'd':	/* debug */
			debug_enabled = TRUE;
			break;
		case 'h':	/* help */
			opt_help = TRUE;
			break;
		case 'n':	/* no autologin */
			opt_nologin = TRUE;
			break;
		case 's':	/* use existing session ID */
			g_free(opt_session_arg);
			opt_session_arg = g_strdup(optarg);
			break;
		case 'v':	/* version */
			opt_version = TRUE;
			break;
		case '?':	/* show terse help */
		default:
			show_usage(argv[0], TRUE);
			return 0;
			break;
		}
	}

	/* show help message */
	if (opt_help) {
		show_usage(argv[0], FALSE);
		return 0;
	}
	/* show version message */
	if (opt_version) {
		printf("gntgaim %s\n", VERSION);
		return 0;
	}

	/* set a user-specified config directory */
	if (opt_config_dir_arg != NULL) {
		gaim_util_set_user_dir(opt_config_dir_arg);
	}

	/*
	 * We're done piddling around with command line arguments.
	 * Fire up this baby.
	 */

	/* Because we don't want debug-messages to show up and corrup the display */
	gaim_debug_set_enabled(debug_enabled);

	gaim_core_set_ui_ops(gnt_core_get_ui_ops());
	gaim_eventloop_set_ui_ops(gnt_eventloop_get_ui_ops());

	path = g_build_filename(gaim_user_dir(), "plugins", NULL);
	gaim_plugins_add_search_path(path);
	g_free(path);

	gaim_plugins_add_search_path(LIBDIR);

	if (!gaim_core_init(GAIM_GNT_UI))
	{
		fprintf(stderr,
				"Initialization of the Gaim core failed. Dumping core.\n"
				"Please report this!\n");
		abort();
	}

	/* TODO: Move blist loading into gaim_blist_init() */
	gaim_set_blist(gaim_blist_new());
	gaim_blist_load();

	/* TODO: Move prefs loading into gaim_prefs_init() */
	gaim_prefs_load();
	gaim_prefs_update_old();

	/* load plugins we had when we quit */
	gaim_plugins_load_saved("/gaim/gnt/plugins/loaded");

	/* TODO: Move pounces loading into gaim_pounces_init() */
	gaim_pounces_load();

	if (opt_nologin)
	{
		/* Set all accounts to "offline" */
		GaimSavedStatus *saved_status;

		/* If we've used this type+message before, lookup the transient status */
		saved_status = gaim_savedstatus_find_transient_by_type_and_message(
							GAIM_STATUS_OFFLINE, NULL);

		/* If this type+message is unique then create a new transient saved status */
		if (saved_status == NULL)
			saved_status = gaim_savedstatus_new(NULL, GAIM_STATUS_OFFLINE);

		/* Set the status for each account */
		gaim_savedstatus_activate(saved_status);
	}
	else
	{
		/* Everything is good to go--sign on already */
		if (!gaim_prefs_get_bool("/core/savedstatus/startup_current_status"))
			gaim_savedstatus_activate(gaim_savedstatus_get_startup());
		gaim_accounts_restore_current_statuses();
	}

	return 1;
}

int main(int argc, char **argv)
{
	/* XXX: Don't puke */
	freopen(".error", "w", stderr);

	/* Initialize the libgaim stuff */
	if (!init_libgaim(argc, argv))
		return 0;

	/* Initialize and run the UI */
	init_gnt_ui();
#ifdef STANDALONE
	gaim_core_quit();
#endif

	return 0;
}

